node/stateSync.go

```diff
@@ -0,0 +1,201 @@
+
+	hashesCh       chan []common.Uint256
+	stateCh        chan [][]byte // Response states channel
+	timeoutReq     chan []common.Uint256
+	failedHashesCh chan []common.Uint256
+	stopChans      []chan struct{}
+
+	peers []*RemoteNode // Set of active peers from which download can proceed
+
+	bytesUncommitted int
+}
+
+// loop is the main event loop of a state trie sync. It is responsible for the
+// assignment of new tasks to peers (including sending it to them) as well as
+// for the processing of inbound data.
+func (s *stateSync) loop() (err error) {
+	defer func() {
+		cerr := s.commit(true)
+		if err == nil {
+			err = cerr
+		}
+		cerr = s.d.database.BatchPut(db.CurrentStateTrie(), s.root.ToArray())
+		if err == nil {
+			err = cerr
+		}
+		if err := s.d.database.BatchCommit(); err != nil {
+			err = cerr
+		}
+	}()
+
+	go s.startWorkerThread()
+
+	go func() {
+		for {
+			log.Info("pending states:", s.sched.Pending())
+			time.Sleep(3 * time.Second)
+			select {
+			case <-s.done:
+				return
+			default:
+			}
+		}
+	}()
+
+	for s.sched.Pending() > 0 {
+		if err = s.commit(false); err != nil {
+			return err
+		}
+
+		for i := 0; i < len(s.peers); i++ {
+			jobs, _ := s.sched.Missing(config.MaxStateReqSize)
+			if len(jobs) == 0 {
+				break
+			}
+			s.hashesCh <- jobs
+		}
+		select {
+		case data := <-s.stateCh:
+			for _, blob := range data {
+				_, err := s.processNodeData(blob)
+				switch err {
+				case nil:
+					s.bytesUncommitted += len(blob)
+				default:
+					log.Info("processNodeData err:", err)
+				}
+			}
+		case <-s.done:
+			return
+		}
+	}
+	return err
+}
+
+func (s *stateSync) commit(force bool) error {
+	if !force && s.bytesUncommitted < config.IdealBatchSize {
+		return nil
+	}
+	s.d.database.NewBatch()
+	if err := s.sched.Commit(s.d.database); err != nil {
+		return err
+	}
+	if err := s.d.database.BatchCommit(); err != nil {
+		return err
+	}
+	s.bytesUncommitted = 0
+	return nil
+}
+
+// processNodeData tries to inject a trie node data blob delivered from a remote
+// peer into the state trie, returning whether anything useful was written or any
+// error occurred.
+func (s *stateSync) processNodeData(blob []byte) (common.Uint256, error) {
+	res := trie.SyncResult{Data: blob}
+	var err error
+	res.Hash, err = common.Uint256ParseFromBytes(s.hasher(blob))
+	if err != nil {
+		return common.EmptyUint256, err
+	}
+	err = s.sched.Process(res)
+	return res.Hash, err
+}
+
+func (s *stateSync) startWorker(ctx context.Context, workerId uint32, p *RemoteNode) {
+	var hashes []common.Uint256
+	for {
+		select {
+		case <-s.stopChans[workerId]:
+			return
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case hashes = <-s.hashesCh:
+		case hashes = <-s.failedHashesCh:
+		case <-s.stopChans[workerId]:
+			return
+		case <-ctx.Done():
+			return
+		}
+
+		resp, err := p.GetStates(hashes)
+
+		if err != nil {
+			s.failedHashesCh <- hashes
+			continue
```

Comment:
We need to distinguish two scenarios: whether we should continue using this worker or not. A few examples:
* if it's timeout, we might give it a few retries before return
* in all other case, we should probably just return